<a href="https://colab.research.google.com/github/RodolfoFerro/ml-facilito/blob/main/notebooks/Modelos_Fundacionales_%26_Hugging_Face_Clase_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Modelos Fundacionales & Hugging Face - Clase 4  🧠

> **Descripción:** Cuaderno de contenidos (IV) sobre Modelos Fundacionales & Hugging Face para el Bootcamp de Machine Learning con Código Facilito, 2023. <br>
> **Autor:** [Rodolfo Ferro](https://github.com/RodolfoFerro) <br>
> **Contacto:** [X](https://twitter.com/rodo_ferro) / [Instagram](https://www.instagram.com/rodo_ferro/)


## Contenido

### Modelos fundacionales

* ¿Qué son los modelos fundacionales?
* Características de los modelos fundacionales
* ¿Por qué es importante el modelado fundacional?
* ¿Cómo funcionan los modelos fundacionales?
* ¿Qué pueden hacer los modelos fundacionales?
* Ejemplos de modelos fundacionales


### Hugging Face - *Transformers*

* a


### Hugging Face – *Fine Tuning* de modelos

* a

## **Modelos fundacionales**

### **¿Qué son los modelos fundacionales?**

Los **modelos fundacionales**, o FMs, son **redes neuronales de aprendizaje profundo** que han transformado el enfoque de los científicos en el machine learning.

*Estos modelos se entrenan con grandes conjuntos.*

> **IDEA:** En lugar de desarrollar IA desde cero, las y los científicos de datos utilizan un modelo fundacional como punto de partida para desarrollar modelos de ML que impulsen aplicaciones nuevas de manera rápida y rentable.


### **Características de los modelos fundacionales**

La **característica distintiva** de los modelos fundacionales es su **capacidad de adaptarse**. Estos modelos pueden realizar diversas tareas con alta precisión, como procesamiento de lenguaje natural, respuesta a preguntas y clasificación robusta de imágenes.

El tamaño y la versatilidad de los modelos fundamentales los distinguen de los modelos convencionales de aprendizaje automático, los cuales típicamente se especializan en tareas particulares, como examinar texto para evaluar opiniones, categorizar imágenes y prever patrones.


### **¿Por qué es importante el modelado fundacional?**

El modelado fundacional tiene el potencial de **revolucionar el ciclo de vida del machine learning**.

Aunque desarrollarlos desde cero es costoso, utilizar modelos pre-entrenados resulta más rápido y económico a largo plazo, impulsando la automatización de diversas aplicaciones.

Una aplicación posible implica la automatización de diversas tareas y procesos, especialmente aquellas que demandan habilidades de razonamiento. Algunos ejemplos de áreas donde los FMs podrían utilizarse:

* Asistencia al cliente
* Traducción de idiomas
* Generación de contenido
* Redacción de textos publicitarios
* Clasificación de imágenes
* Creación y edición de imágenes de alta resolución
* Extracción de documentos
* Sector de la salud



### **¿Cómo funcionan los modelos fundacionales?**

Los modelos fundacionales son una forma de **IA generativa**, operan mediante complejas redes neuronales, como las generativas antagónicas/adversariales (GANs), los transformadores (Transformers) y los autoencoders variacionales.

> → Ellos generan resultados a partir de una o más entradas (indicaciones) en forma de instrucciones en lenguaje humano.


### **¿Qué pueden hacer los modelos fundacionales?**

A pesar de ser pre-entrenados, los modelos fundacionales pueden continuar aprendiendo de nuevas entradas, realizando tareas como:

* Procesamiento del lenguaje natural
* Comprensión visual
* Generación de código
* Voz a texto
* Participación centrada en las personas


### **Ejemplos de modelos fundacionales**

La oferta de modelos fundacionales ha experimentado un rápido crecimiento en tamaño y cantidad en el mercado. Algunos ejemplos notables incluyen:

* BERT
* GPTs
* Stable Diffusion
* BLOOM
* Gemini…
* …y la plataforma Hugging Face.


## **Hugging Face - Transformers**

### **Hugging Face**

<center>
    <img src="https://huggingface.co/front/assets/huggingface_logo-noborder.svg" width="30%">
</center>


[Hugging Face](https://huggingface.co/) es una plataforma que ofrece herramientas open-source para la creación e implementación de modelos de ML. Funciona como un punto de encuentro comunitario donde las y los desarrolladores tienen la posibilidad de intercambiar y explorar modelos y conjuntos de datos.

**La plataforma ofrece acceso público a casi 200,000 modelos y 30,000 conjuntos de datos.**

> La membresía individual es de acceso gratuito, mientras que las
suscripciones de pago brindan niveles de acceso más avanzados.


### **¿Qué es NLP?**

El **Procesamiento del Lenguaje Natural** (PLN/NLP) es un campo de la lingüística y el aprendizaje automático que se centra en comprender todo lo relacionado con el lenguaje humano. Su objetivo no es solo entender palabras individualmente, sino comprender el contexto en el que se utilizan.

La siguiente es una lista de tareas comunes de procesamiento del lenguaje natural (NLP), con algunos ejemplos de cada una:

* **Clasificación de oraciones completas:** Obtener el sentimiento de una reseña, detectar si un correo electrónico es spam, determinar si una oración es gramaticalmente correcta o si dos oraciones están lógicamente relacionadas o no.
* **Clasificación de cada palabra en una oración:** Identificar los componentes gramaticales de una oración (sustantivo, verbo, adjetivo) o las entidades nombradas (persona, ubicación, organización).
* **Generación de contenido de texto:** Completar un estímulo con texto generado automáticamente, llenar los espacios en blanco en un texto con palabras enmascaradas.
* **Extracción de una respuesta de un texto:** Dada una pregunta y un contexto, extraer la respuesta a la pregunta basándose en la información proporcionada en el contexto.
Generación de una nueva oración a partir de un texto de entrada: Traducir un texto a otro idioma, resumir un texto.

#### **¿Por qué es difícil?**

Las computadoras no manejan ni procesan la información de la misma manera que lo hacemos los humanos.

Para los modelos de ML, estas tareas son complicadas. Es necesario procesar el texto de manera que permita al modelo aprender de él.

Dada la complejidad del lenguaje, es crucial reflexionar sobre cómo llevar a cabo este procesamiento. Se ha llevado a cabo mucha investigación sobre cómo representar el texto.

### ***Transformers***

Los modelos **Transformer** se emplean para abordar diversas tareas de procesamiento del lenguaje natural.

Algunas empresas y organizaciones aprovechan Hugging Face y los modelos *Transformer*, devolviendo también a la comunidad al compartier sus propios modelos.

La biblioteca [🤗 Transformers](https://github.com/huggingface/transformers) brinda la funcionalidad necesaria para crear y utilizar estos modelos compartidos.

Además, el [Model Hub](https://huggingface.co/models) alberga miles de modelos preentrenados que están disponibles para descargar y utilizar por cualquier persona.

Para comenzar a usar transformers, la biblioteca Transformers cuenta con el elemento más básico, que es la función `pipeline()`.

Esta función conecta un modelo con sus pasos de preprocesamiento y postprocesamiento necesarios, lo que nos permite ingresar directamente cualquier texto y obtener una respuesta comprensible.

In [ ]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis")
classifier("¡Me encanta el bootcamp de ML de Código Facilito!")

In [ ]:
classifier([
    "Feliz navidad para todos ustedes, jingle bells, jingle bells, jingle...",
    "Hoy no me gusta cómo está el clima."
])

De manera predeterminada, este `pipeline` selecciona un modelo pre-entrenado específico que ha sido ajustado para analizar sentimientos. El modelo se descarga y almacena en caché al crear el objeto clasificador.

> **Nota:** Si vovlemos a ejecutar el comando, se utilizará el modelo almacenado en caché, evitando la necesidad de volver a descargarlo.

Cuando ingresamos un texto en este pipeline, se llevan a cabo tres pasos esenciales:

1. El texto se transforma previamente a un formato que el modelo pueda comprender.
2. Las entradas preprocesadas se suministran al modelo.
3. Las predicciones del modelo se procesan posteriormente para que tengan sentido.

Algunos de los pipelines que están disponibles en la actualidad incluyen:

* `feature-extraction` (obtención de la representación vectorial de un texto)
* `question-answering`
* `sentiment-analysis` (análisis de sentimientos)
* `summarization` (resumen)
* `text-generation` (generación de texto)
* `translation` (traducción)
* `zero-shot-classification` (clasificación sin etiquetas previas)

### **Uso de Transormers**

#### **Generación de texto**

In [ ]:
from transformers import pipeline

generator = pipeline("text-generation")

In [ ]:
generator("En este bootcamp sobre Machine Learning de Código Facilito aprenderás")

In [ ]:
generator("In this bootcamp about ML, you'll learn")

Podemos utilizar algún modelo específico, simplemente yendo al [Model Hub](https://huggingface.co/models) y dando click en el tag correspondiente para desplegar modelos asociados a la tarea.

In [ ]:
generator = pipeline("text-generation", model="distilgpt2")

In [ ]:
generator(
    "In this bootcamp about Machine Learning, you'll learn",
    max_length=30,
    num_return_sequences=2,
)

#### **Resumen de contenido**

In [ ]:
from transformers import pipeline

summarizer = pipeline("summarization")
summarizer(
    """
    La característica distintiva de los modelos fundacionales es su
    capacidad de adaptarse. Estos modelos pueden realizar diversas
    tareas con alta precisión, como procesamiento de lenguaje natural,
    respuesta a preguntas y clasificación robusta de imágenes.

    El tamaño y la versatilidad de los modelos fundamentales los
    distinguen de los modelos convencionales de aprendizaje automático,
    los cuales típicamente se especializan en tareas particulares, como
    examinar texto para evaluar opiniones, categorizar imágenes y prever
    patrones.
"""
)

#### **Respuestas a preguntas**

In [ ]:
from transformers import pipeline

question_answerer = pipeline("question-answering")
question_answerer(
    question="¿Cómo me llamo?",
    context="Me llamo Rodo. Nací en Guanajuato, México.",
)

In [ ]:
question_answerer(
    question="Where am I from?",
    context="My name is Rodo. I was born in Guanajuato, México.",
)

#### **Otras tareas**

Puedes consultar otras tareas de ejemplo en esta [referencia](https://huggingface.co/learn/nlp-course/chapter1/3).

### **¿Cómo funcionan?**

Podríamos decir que en general los Transformers se pueden agrupar en tres categorías:

* Estilo GPT (también llamados modelos auto-regresivos Transformer)
* Estilo BERT (también llamados modelos auto-codificadores Transformer)
* Estilo BART/T5 (también llamados modelos Transformer de secuencia a secuencia)

Todos los modelos Transformer mencionados anteriormente (GPT, BERT, BART, T5, etc.) han sido entrenados como modelos de lenguaje. Esto significa que han sido entrenados en grandes cantidades de texto sin procesar de manera auto supervisada.

> El aprendizaje auto supervisado es un tipo de entrenamiento en el que el objetivo se calcula automáticamente a partir de las entradas del modelo. ¡Esto significa que no se necesita la intervención humana para etiquetar los datos!

Este tipo de modelos desarrollan una comprensión estadística del lenguaje en el que ha sido entrenado, pero no es muy útil para tareas prácticas específicas. Debido a esto, el modelo preentrenado general pasa por un proceso de transfer learning. Durante este proceso, el modelo se ajusta finamente de manera supervisada, es decir, utilizando etiquetas anotadas por humanos, en una tarea específica.

Un ejemplo de tarea es predecir la siguiente palabra en una oración después de haber leído las `n` palabras anteriores. Esto se llama modelado de lenguaje causal porque la salida depende de las entradas pasadas y presentes, pero no de las futuras.

<center>
    <img src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter1/causal_modeling.svg" width="80%">
</center>

#### **Arquitectura general**

El modelo está principalmente compuesto por dos bloques:

* **Encoder:** El codificador recibe una entrada y construye una representación de ella (sus características). Esto significa que el modelo está optimizado para adquirir comprensión a partir de la entrada.

* **Decoder:** El decodificador utiliza la representación del codificador (características) junto con otras entradas para generar una secuencia objetivo. Esto significa que el modelo está optimizado para generar salidas.

<center>
    <img src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter1/transformers_blocks.svg" width="60%">
</center>

Cada una de estas partes puede utilizarse de forma independiente según la tarea:

* **Modelos sólo de codificación:** Útiles para tareas que requieren comprensión de la entrada, como la clasificación de oraciones y el reconocimiento de entidades nombradas.
* **Modelos sólo de decodificación:** Apropiados para tareas generativas como la generación de texto.
* **Modelos de codificación-decodificación o modelos de secuencia a secuencia:** Eficientes para tareas generativas que requieren una entrada, como la traducción o la síntesis.

#### **Capas de atención**

Una característica clave de los modelos Transformer es que están construidos con capas especiales llamadas **capas de atención**. De hecho, el título del artículo que introduce la arquitectura Transformer es ["Attention Is All You Need"](https://arxiv.org/abs/1706.03762) ("La Atención es Todo lo que Necesitas"). Esta capa indicará al modelo que preste atención específica a ciertas palabras en la oración que le pasamos (y más o menos ignore las demás) al tratar con la representación de cada palabra.

> Para contextualizar esto, considera la tarea de traducir texto de inglés a francés. Dada la entrada "You like this course" ("Te gusta este curso"), un modelo de traducción también necesitará prestar atención a la palabra adyacente "You" para obtener la traducción adecuada de la palabra "like", porque en francés el verbo "like" se conjuga de manera diferente según el sujeto. El resto de la oración, sin embargo, no es útil para la traducción de esa palabra. De manera similar, al traducir "this" ("esto" o "esta"), el modelo también deberá prestar atención a la palabra "course" ("curso"), porque "this" se traduce de manera diferente según si el sustantivo asociado es masculino o femenino. Nuevamente, las otras palabras en la oración no importarán para la traducción de "this". Con oraciones más complejas (y reglas gramaticales más complejas), el modelo necesitaría prestar especial atención a palabras que puedan aparecer más lejos en la oración para traducir correctamente cada palabra.

El mismo concepto se aplica a cualquier tarea asociada con el lenguaje natural: una palabra por sí sola tiene un significado, pero ese significado se ve profundamente afectado por el contexto, que puede ser cualquier otra palabra (o palabras) antes o después de la palabra que se está estudiando.

#### **La arquitectura original**

* La arquitectura Transformer fue diseñada originalmente para la traducción. Durante el entrenamiento, el codificador recibe entradas (oraciones) en un cierto idioma, mientras que el decodificador recibe las mismas oraciones en el idioma objetivo deseado.
* En el codificador, las capas de atención pueden usar todas las palabras en una oración (ya que, como acabamos de ver, la traducción de una palabra dada puede depender de lo que haya después y antes en la oración).
* Sin embargo, el decodificador trabaja de manera secuencial y solo puede prestar atención a las palabras en la oración que ya ha traducido (es decir, solo las palabras antes de la palabra que se está generando en ese momento).

Para acelerar las cosas durante el entrenamiento (cuando el modelo tiene acceso a las oraciones objetivo), se le proporciona al decodificador todo el objetivo, pero no se le permite usar palabras futuras. Por ejemplo, al intentar predecir la cuarta palabra, la capa de atención solo tendrá acceso a las palabras en las posiciones 1 a 3.

La arquitectura original de Transformer se veía así, con el codificador a la izquierda y el decodificador a la derecha:

<center>
    <img src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter1/transformers.svg" width="90%">
</center>

Observa que la primera capa de atención en un bloque de decodificador presta atención a todas las entradas (pasadas) al decodificador, pero la segunda capa de atención utiliza la salida del codificador. Por lo tanto, puede acceder a toda la oración de entrada para predecir mejor la palabra actual. Esto es muy útil, ya que diferentes idiomas pueden tener reglas gramaticales que colocan las palabras en diferentes órdenes, o algún contexto proporcionado más tarde en la oración puede ser útil para determinar la mejor traducción de una palabra dada.

La máscara de atención también se puede utilizar en el codificador/decodificador para evitar que el modelo preste atención a algunas palabras especiales, como la palabra de relleno especial utilizada para igualar la longitud de todas las entradas al agrupar oraciones.

#### **Entonces...**

Estos elementos te resumen un poco el funcionamiento en alto nivel sobre este tipo de modelos:

- Encoding posicional
- Atención
    - Atención propia (ahora con contexto sobre las palabras alrededor de la observada)

Te recomiendo mucho ver estos videos:


*   [¿Qué es un TRANSFORMER? La Red Neuronal que lo cambió TODO!](https://www.youtube.com/watch?v=aL-EmKuB078)
*   [Transformers, explained: Understand the model behind GPT, BERT, and T5](https://www.youtube.com/watch?v=SZorAJ4I-sA)
* [Illustrated Guide to Transformers Neural Network: A step by step explanation](https://www.youtube.com/watch?v=4Bdc55j80l8)



### **Modelos y tokenizers**

Como mencionamos, `pipeline` agrupa tres pasos: preprocesamiento de entradas, paso de las entradas a través del modelo y posprocesamiento.


<center>
    <img src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/full_nlp_pipeline.svg" width="90%">
</center>

#### **Preprocesamiento con un tokenizer**

Al igual que otras redes neuronales, los modelos Transformer no pueden procesar texto directamente, por lo que el primer paso de nuestro flujo (pipeline) es convertir las entradas de texto en números comprensibles para el modelo. Para ello, utilizamos un tokenizador, que se encargará de:

1. Dividir la entrada en palabras, subpalabras o símbolos (como la puntuación) llamados tokens.
2. Asignar un número entero a cada token.
3. Agregar entradas adicionales que puedan ser útiles para el modelo.

Todo este preprocesamiento debe realizarse de la misma manera que cuando el modelo fue pre-entrenado.


In [17]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

In [18]:
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

In [30]:
raw_inputs = [
    "I hate when my neighbor does not feed her cat.",
    "I enjoy coding in Python.",
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)

{'input_ids': tensor([[  101,  1045,  5223,  2043,  2026, 11429,  2515,  2025,  5438,  2014,
          4937,  1012,   102],
        [  101,  1045,  5959, 16861,  1999, 18750,  1012,   102,     0,     0,
             0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0]])}


La salida en sí es un diccionario que contiene dos claves, `input_ids` y `attention_mask`. `input_ids` contiene dos tensores (de PyTorch) con números enteros (una para cada oración) que son los identificadores únicos de los tokens en cada oración.

En cuanto a `attention_mask`, una máscara de atención es una máscara binaria que designa qué tokens deben ser atendidos (asignándoles pesos no nulos) y cuáles deben ser ignorados (asignándoles pesos nulos). Al aplicar esta máscara, el modelo puede atender selectivamente a tokens específicos mientras ignora otros.

#### **Procesamiento con el modelo**

Ahora podemos descargar un modelo pre-entrenado de la misma manera que lo hicimos con nuestro tokenizador. Para ello, podemos usar el objeto `AutoModel` que también tiene un método `from_pretrained()`:

In [31]:
from transformers import AutoModel

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

In [32]:
model = AutoModel.from_pretrained(checkpoint)

Esta arquitectura contiene solo el módulo base del Transformer: dados algunos inputs, produce lo que llamaremos estados ocultos, también conocidos como características. Para cada input del modelo, recuperaremos un **vector de alta dimensionalidad** que representa la comprensión contextual de ese input por parte del modelo Transformer.

In [33]:
outputs = model(**inputs)
print(outputs.last_hidden_state.shape)

torch.Size([2, 13, 768])


##### **¿Un vector de alta dimensionalidad?**

El vector producido por el módulo Transformer suele ser grande. Generalmente tiene tres dimensiones:

- **Tamaño del batch:** El número de secuencias procesadas a la vez (2 en nuestro ejemplo).
- **Longitud de la secuencia:** La longitud de la representación numérica de la secuencia (13 en este ejemplo).
- **Tamaño oculto:** La dimensión del vector de cada input del modelo.
Se dice que es "de alta dimensionalidad" debido al último valor. El tamaño oculto puede ser muy grande (768 es común para modelos más pequeños, y en modelos más grandes esto puede alcanzar 3072 o más).

##### **Cabezas del modelo**

Las cabezas del modelo toman el vector de alta dimensionalidad de estados ocultos como input y los proyectan a una dimensión diferente. Por lo general, están compuestas por una o unas pocas capas lineales:

<center>
    <img src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/transformer_and_head.svg" width="90%">
</center>

La salida del modelo Transformer se envía directamente a la cabeza del modelo para ser procesada.

El modelo está representado por su capa de embeddings y las capas subsiguientes. La capa de embeddings convierte cada ID de input en el input tokenizado en un vector que representa el token asociado. Las capas subsiguientes manipulan esos vectores utilizando el mecanismo de atención para producir la representación final de las oraciones.

Hay muchas arquitecturas diferentes disponibles en Transformers, cada una diseñada para abordar una tarea específica. Aquí hay una lista no exhaustiva:

* Model
* ForCausalLM
* ForMaskedLM
* ForMultipleChoice
* ForQuestionAnswering
* ForSequenceClassification
* ForTokenClassification

Para nuestro ejemplo, necesitaremos un modelo con una cabeza de clasificación de secuencia (para poder clasificar las oraciones como positivas o negativas). Por lo tanto, en realidad, no usaremos la clase `AutoModel`, sino `AutoModelForSequenceClassification`:

In [34]:
from transformers import AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

In [35]:
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)

In [36]:
outputs = model(**inputs)
print(outputs.logits.shape)

torch.Size([2, 2])


> **Pregunta:** ¿Por qué es de 2x2?

#### **Postprocesaiento de la salida**

In [37]:
print(outputs.logits)

tensor([[ 2.7505, -2.3348],
        [-2.4761,  2.5623]], grad_fn=<AddmmBackward0>)


In [38]:
import torch

predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)
print(predictions)

tensor([[0.9939, 0.0061],
        [0.0064, 0.9936]], grad_fn=<SoftmaxBackward0>)


In [39]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}

### **Siguientes pasos...**

- Preparación de los datos
- Fine-tuning de modelos:
    - Con la API Trainer
    - Con Keras

Referencias: https://huggingface.co/learn/nlp-course/en/chapter0/1?fw=pt

--------

> Contenido creado por **Rodolfo Ferro**, 2023. <br>
> Puedes contactarme a través de Insta ([@rodo_ferro](https://www.instagram.com/rodo_ferro/)) o X ([@rodo_ferro](https://twitter.com/rodo_ferro)).